### ISSUE TITLE
BUG: DataFrame and Series constructor inconsistent in handling of Datetimelike object index

### ISSUE THREAD

**Issue by phofl** [pandas-dev/pandas#39307]
- [x] I have checked that this issue has not already been reported. (Looked through and could not find anything)

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import pandas as pd
idx = pd.Index([pd.Timestamp("2011-01-01")], dtype=object)
ser = pd.Series([1], index=idx)
df = pd.DataFrame([1], index=idx)
ser.index
Out[4]: DatetimeIndex(['2011-01-01'], dtype='datetime64[ns]', freq=None)
df.index
Out[5]: Index([2011-01-01 00:00:00], dtype='object')

```

#### Problem description

The behavior should be consistent, but not sure what to expected here. The inference is a nice feature although.
The current behavior leads to weird situations with indexing:

```
ser.loc["2011"]
Out[7]: 
2011-01-01    1
dtype: int64
```
works while
```
df.loc["2011"]
Traceback (most recent call last):
  File "/home/developer/PycharmProjects/pandas/pandas/core/indexes/base.py", line 3225, in get_loc
    return self._engine.get_loc(casted_key)
  File "pandas/_libs/index.pyx", line 70, in pandas._libs.index.IndexEngine.get_loc
    cpdef get_loc(self, object val):
  File "pandas/_libs/index.pyx", line 101, in pandas._libs.index.IndexEngine.get_loc
    return self.mapping.get_item(val)
  File "pandas/_libs/hashtable_class_helper.pxi", line 5038, in pandas._libs.hashtable.PyObjectHashTable.get_item
    cpdef get_item(self, object val):
  File "pandas/_libs/hashtable_class_helper.pxi", line 5046, in pandas._libs.hashtable.PyObjectHashTable.get_item
    raise KeyError(val)
KeyError: '2011'
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/home/developer/anaconda3/envs/omi_reports/pandas-dev/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3417, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-8-59984db8d56a>", line 1, in <module>
    df.loc["2011"]
  File "/home/developer/PycharmProjects/pandas/pandas/core/indexing.py", line 902, in __getitem__
    return self._getitem_axis(maybe_callable, axis=axis)
  File "/home/developer/PycharmProjects/pandas/pandas/core/indexing.py", line 1128, in _getitem_axis
    return self._get_label(key, axis=axis)
  File "/home/developer/PycharmProjects/pandas/pandas/core/indexing.py", line 1077, in _get_label
    return self.obj.xs(label, axis=axis)
  File "/home/developer/PycharmProjects/pandas/pandas/core/generic.py", line 3752, in xs
    loc = index.get_loc(key)
  File "/home/developer/PycharmProjects/pandas/pandas/core/indexes/base.py", line 3227, in get_loc
    raise KeyError(key) from err
KeyError: '2011'
```

#### Expected Output

Bot returning the same index dtype

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : ff628b17ccfcaf14ceefeb438aafa56d3849937c
python           : 3.8.6.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.8.0-38-generic
Version          : #43~20.04.1-Ubuntu SMP Tue Jan 12 16:39:47 UTC 2021
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8
pandas           : 1.3.0.dev0+483.gff628b17cc.dirty
numpy            : 1.21.0.dev0+438.gbf4ac1f81
pytz             : 2020.1
dateutil         : 2.8.1
pip              : 20.2.3
setuptools       : 49.6.0.post20201009
Cython           : 0.29.21
pytest           : 6.1.1
hypothesis       : 5.37.1
sphinx           : 3.4.3
blosc            : None
feather          : None
xlsxwriter       : 1.3.7
lxml.etree       : 4.5.2
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.18.1
pandas_datareader: None
bs4              : 4.9.3
bottleneck       : 1.3.2
fsspec           : 0.8.3
fastparquet      : 0.4.1
gcsfs            : 0.7.1
matplotlib       : 3.3.2
numexpr          : 2.7.1
odfpy            : None
openpyxl         : 3.0.5
pandas_gbq       : None
pyarrow          : 2.0.0
pyxlsb           : None
s3fs             : 0.4.2
scipy            : 1.7.0.dev0+5e9eb01
sqlalchemy       : 1.3.20
tables           : 3.6.1
tabulate         : 0.8.7
xarray           : 0.16.1
xlrd             : 1.2.0
xlwt             : 1.3.0
numba            : 0.51.2

</details>

**Comment by jbrockmendel** [comment]
IIRC we deprecated the series behavior but then un-deprecated just before 1.2.0 bc it was causing futurewarnings that we hadn’t found a good way to suppress.  +1 for re-deprecating

**Comment by jbrockmendel** [comment]
#38679

**Comment by simonjayhawkins** [comment]
> +1 for re-deprecating

I think we accepted that we would defer the deprecation to 1.3 (and fix the warnings issue)

We could either
1) undo the revert (#38679) and reopen #38182
2) open an issue for the deprecation (milestoned 1.3) so it is not forgotten ( and only merge an associated PR when happy)

if we do 1, we could end up reverting again if the issue not resolved before 1.3 release.

cc @jorisvandenbossche
